### ISSUE TITLE
benchmark.py: state leaks across examples, and MCC is measured on the training stream

### ISSUE THREAD

**Issue by mar-i0** [jrz97619761/test-model-thing#4]
Two independent problems in `benchmark.py` that I think make the reported CoLA number not measurable.

## 1. Internal state carries across examples

```python
for i, (bytes, label) in enumerate(data):
    final = None
    for b in bytes:
        x = model.encoder(mx.array(b))
        for j, layer in enumerate(model.layers):
            x, state, _ = layer(x, dummies[j])
            layer.states = mx.stop_gradient(state)
```

`layer.states` is never reset between examples, so the representation of sentence `N` is a function of sentences `1..N-1` as well. CoLA sentences are independent, so this is leakage, and it makes the probe's input depend on the shuffle order. A `model.reset()` (zeroing `layer.states`, `layer.decaytrace` and `encoder.embedtrace`) before each example fixes it.

Worth noting this is only a bug *for this benchmark*. Persistent state is the whole point of the architecture — it just isn't what you want when scoring independent sentences.

## 2. MCC is accumulated over the training stream

```python
predicted_class = mx.argmax(choice).item()
if predicted_class == 1 and label == 1: tp += 1
...
score = mcc(tp, tn, fp, fn)
```

`choice` here is the head's output on the example it is *about to train on*, and the counters accumulate from the first example of the epoch onwards, across all three epochs. So the reported MCC mixes the head's early untrained predictions with later ones, on data it is being fit to, with no held-out split. That number can't distinguish a backbone that carries linguistic structure from one that memorises — which is exactly the question the benchmark exists to answer.

What I'd suggest: hold out ~20% up front, train the head on the rest, and report MCC on the held-out part only, recomputed per epoch rather than accumulated. It's also much faster to encode every sentence through the frozen backbone once and reuse the vectors across epochs, since the backbone doesn't change.

## Minor

- `for i, (bytes, label) in enumerate(data)` shadows the builtin `bytes`.
- The final `print` uses `i` and `score`, which are undefined if `data` is empty (e.g. wrong path, or the `len(parts) == 4` filter drops every row) — you get a `NameError` instead of a useful message.
- `model.freeze()` is called but the backbone still builds a graph per byte; wrapping the encode loop in a no-grad context is a large speedup.

I have a reworked version in my PyTorch port if it's useful: https://github.com/mar-i0/lain

---

*Found with [Claude Code](https://claude.com/claude-code) while porting the model to PyTorch. The port, with everything needed to reproduce the numbers above, is at https://github.com/mar-i0/lain.*

**Comment by jrz97619761** [comment]
Internal state issue and the minor issues are fixed, although I haven't implemented the benchmark optimizations yet. Not sure about your point that the MCC is accumulated, it resets every epoch no?

**Comment by jrz97619761** [comment]
Should be all fixed now. I'll close this, but feel free to reopen if you want
